staging/src/k8s.io/client-go/plugin/pkg/client/auth/exec/exec.go

```diff
@@ -106,6 +108,31 @@ func (c *cache) put(s string, a *Authenticator) *Authenticator {
 	return a
 }
 
+// returns the shortest certificate time to live (TTL) (derived from the
+// oldest .NotAfter) among all Authenticator objects within the cache. Useful
+// for watching if any certificates are nearing expiration. If cache contains
+// no Authenticators or no Authenticators contain a cached TLS certificate
+// this function returns nil.
+func (c *cache) shortestTTL(now func() time.Time) *time.Duration {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	var shortest *time.Duration
+	for _, a := range c.m {
+		if cert := a.leafCertificate(); cert != nil {
```

Comment:
rather than looping back from the global cache to authenticators, it would be cleaner for authenticators to push expirys into something tracking that information.

Something along the lines of:

```go
type certificateExpirationTracker struct {
	mu       sync.Mutex
	m        map[*Authenticator]time.Time
	earliest time.Time
}

var certificateExpirations = &certificateExpirationTracker{m: map[*Authenticator]time.Time{}}

// set stores the given expiration time, updates earliest, and returns the reports the earliest expiration time.
// If no Authenticators have reported a certificate expiration, this function returns a zero time.
func (c *certificateExpirationTracker) set(a *Authenticator, t time.Time) time.Time {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.m[a] = t

	// update earliest
	earliest := time.Time{}
	for _, t := range c.m {
		if t.IsZero() {
			continue
		}
		if earliest.IsZero() || earliest.After(t) {
			earliest = t
		}
	}
	c.earliest = earliest
	return earliest
}

// report reports and returns the ttl to the earliest reported expiration time.
// If no Authenticators have reported a certificate expiration, this function returns nil.
func (c *certificateExpirationTracker) report(now func() time.Time) *time.Duration {
	c.mu.Lock()
	defer c.mu.Unlock()
	if c.earliest.IsZero() {
		metrics.ClientCertTTL.Set(nil)
		return nil
	} else {
		ttl := c.earliest.Sub(now())
		metrics.ClientCertTTL.Set(&ttl)
		return &ttl
	}
}
```

used from `refreshCredsLocked` like this:
```go
	expiry := time.Time{}
	if a.cachedCreds.cert != nil {
		expiry = a.cachedCreds.cert.Leaf.NotAfter
	}
	certificateExpirations.set(a, expiry)
	certificateExpirations.report(time.Now)
```

Reply:
Done. Except I moved to `.report(...)` to `getCreds()`